src/libre/re_analysis.c

```diff
@@ -5,78 +5,645 @@
+
+	case AST_EXPR_ANCHOR:
+		return 0;
+
+	case AST_EXPR_GROUP:
+		if (thud) { return 1; } /* FIXME */
+		return always_consumes_input(n->u.group.e, thud);
+
+	default:
+		return 0;
+	}
+}
+
+static int
+is_start_anchor(const struct ast_expr *n) {
+	return n->t == AST_EXPR_ANCHOR && n->u.anchor.t == RE_AST_ANCHOR_START;
+}
+
+static int
+is_end_anchor(const struct ast_expr *n) {
+	return n->t == AST_EXPR_ANCHOR && n->u.anchor.t == RE_AST_ANCHOR_END;
+}
+
+static enum re_analysis_res
+analysis_iter_anchoring(struct anchoring_env *env, struct ast_expr *n)
+{
+	enum re_analysis_res res;
+	/* Flag the overall RE as unsatisfiable if there are anchors
+	 * that cannot be simultaneously satisfied. */
+	
+	switch (n->t) {
+	case AST_EXPR_EMPTY:
+	case AST_EXPR_FLAGS:
+		break;
+
+	case AST_EXPR_ANCHOR:
+		if (n->u.anchor.t == RE_AST_ANCHOR_START) {
+			/* If it's not possible to get here without consuming
+			 * any input and there's a start anchor, the regex is
+			 * inherently unsatisfiable. */
+
+			if (env->past_any_consuming) {
+				set_flags(n, RE_AST_FLAG_UNSATISFIABLE);
+				/* Start anchors that are not in a FIRST position
+				 * almost perfectly match unsatisfiable start
+				 * anchors, with the exception of multiple start
+				 * anchors in a row (e.g. `^^ab$`). We can't just
+				 * treat them as nullable either, because it messes
+				 * up the results for linking. */
+				assert(0 == (n->flags & RE_AST_FLAG_FIRST_STATE));
+				return RE_ANALYSIS_UNSATISFIABLE;
+			}
+		} else if (n->u.anchor.t == RE_AST_ANCHOR_END) {
+			if (env->followed_by_consuming) {
+				/* See above: This is a near perfect subset of
+				 * unsatisfiable cases, but fails to handle
+				 * redundant $ anchors correctly (including
+				 * ones with arbitrary nesting). */
+				assert(0 == (n->flags & RE_AST_FLAG_LAST_STATE));
+				set_flags(n, RE_AST_FLAG_UNSATISFIABLE);
+				return RE_ANALYSIS_UNSATISFIABLE;
+			}
+		} else {
+			assert(0);
+		}
+		break;
+
+	/* These are the types that actually consume input. */
+	case AST_EXPR_LITERAL:
+	case AST_EXPR_ANY:
+	case AST_EXPR_CHAR_CLASS:
+		if (!is_nullable(n)) {
+			env->past_any_consuming = 1;
+		}
+		break;
+
+	case AST_EXPR_CONCAT_N:
+	{
+		size_t i, j;
+		char bak_consuming_after = 0;
+
+		for (i = 0; i < n->u.concat_n.count; i++) {
+			struct ast_expr *child = n->u.concat_n.n[i];
+#if LOG_CONCAT_FLAGS
+			fprintf(stderr, "%s: %p: %lu: %p -- past_any %d\n",
+			    __func__, (void *)n, i, (void *)child,
+			    env->past_any_consuming);
+#endif
+
+			/* Before recursing, save and restore whether
+			 * this is followed by anything which always consumes input.
+			 *
+			 * TODO: instead of doing a backward sweep for each node,
+			 * do a reverse falsification step as a separate pass? */
+			bak_consuming_after = env->followed_by_consuming;
+
+			/* If we don't already know whether this is being followed
+			 * by something that always consumes input, check */
+			if (!env->followed_by_consuming) {
```

Comment:
The tricky part here is that it can't just save whether there's an end anchor and reject the regex if it's followed by other nodes which must consume input -- the end anchor may be in an alt, and cases like `^abc(d|$)e$` should just prune the unsatisfiable anchor and still match otherwise.

Currently it's scanning more than it needs to (it could probably do it once and cache), but it seems to need a 'followed by consuming nodes' flag passed into the subtree.